rmw_fastrtps_cpp/src/rmw_init.cpp

```diff
@@ -12,64 +13,69 @@
+#include "rmw/init_options.h"
+#include "rmw/publisher_options.h"
 #include "rmw/rmw.h"
 
+#include "rmw_dds_common/context.hpp"
+#include "rmw_dds_common/msg/participant_entities_info.hpp"
+
+#include "rmw_fastrtps_shared_cpp/custom_participant_info.hpp"
+#include "rmw_fastrtps_shared_cpp/participant.hpp"
+#include "rmw_fastrtps_shared_cpp/publisher.hpp"
+#include "rmw_fastrtps_shared_cpp/rmw_context_impl.hpp"
+#include "rmw_fastrtps_shared_cpp/rmw_init.hpp"
+#include "rmw_fastrtps_shared_cpp/subscription.hpp"
+
+#include "rosidl_typesupport_cpp/message_type_support.hpp"
+
 #include "rmw_fastrtps_cpp/identifier.hpp"
+#include "rmw_fastrtps_cpp/publisher.hpp"
+#include "rmw_fastrtps_cpp/subscription.hpp"
 
 extern "C"
 {
 rmw_ret_t
 rmw_init_options_init(rmw_init_options_t * init_options, rcutils_allocator_t allocator)
 {
-  RMW_CHECK_ARGUMENT_FOR_NULL(init_options, RMW_RET_INVALID_ARGUMENT);
-  RCUTILS_CHECK_ALLOCATOR(&allocator, return RMW_RET_INVALID_ARGUMENT);
-  if (NULL != init_options->implementation_identifier) {
-    RMW_SET_ERROR_MSG("expected zero-initialized init_options");
-    return RMW_RET_INVALID_ARGUMENT;
-  }
-  init_options->instance_id = 0;
-  init_options->implementation_identifier = eprosima_fastrtps_identifier;
-  init_options->allocator = allocator;
-  init_options->impl = nullptr;
-  return RMW_RET_OK;
+  return rmw_fastrtps_shared_cpp::rmw_init_options_init(
+    eprosima_fastrtps_identifier, init_options, allocator);
 }
 
 rmw_ret_t
 rmw_init_options_copy(const rmw_init_options_t * src, rmw_init_options_t * dst)
 {
-  RMW_CHECK_ARGUMENT_FOR_NULL(src, RMW_RET_INVALID_ARGUMENT);
-  RMW_CHECK_ARGUMENT_FOR_NULL(dst, RMW_RET_INVALID_ARGUMENT);
-  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
-    src,
-    src->implementation_identifier,
-    eprosima_fastrtps_identifier,
-    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
-  if (NULL != dst->implementation_identifier) {
-    RMW_SET_ERROR_MSG("expected zero-initialized dst");
-    return RMW_RET_INVALID_ARGUMENT;
-  }
-  *dst = *src;
-  return RMW_RET_OK;
+  return rmw_fastrtps_shared_cpp::rmw_init_options_copy(
+    eprosima_fastrtps_identifier, src, dst);
 }
 
 rmw_ret_t
 rmw_init_options_fini(rmw_init_options_t * init_options)
 {
-  RMW_CHECK_ARGUMENT_FOR_NULL(init_options, RMW_RET_INVALID_ARGUMENT);
-  RCUTILS_CHECK_ALLOCATOR(&(init_options->allocator), return RMW_RET_INVALID_ARGUMENT);
-  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
-    init_options,
-    init_options->implementation_identifier,
-    eprosima_fastrtps_identifier,
-    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
-  *init_options = rmw_get_zero_initialized_init_options();
-  return RMW_RET_OK;
+  return rmw_fastrtps_shared_cpp::rmw_init_options_fini(
+    eprosima_fastrtps_identifier, init_options);
 }
 
+using rmw_dds_common::msg::ParticipantEntitiesInfo;
+
 rmw_ret_t
 rmw_init(const rmw_init_options_t * options, rmw_context_t * context)
 {
+  std::unique_ptr<rmw_context_t, void (*)(rmw_context_t *)> clean_when_fail(
```

Comment:
@ivanpauno meta, not for now: this seems like a duplicate of the one in `rmw_fastrtps_dynamic_cpp`

Reply:
Yes, the problem is that I can't move it to `rmw_fastrtps_cpp_shared`, because it uses functions defined in `rmw_fastrtps_cpp`.

Reply:
See https://github.com/ros2/rmw_fastrtps/pull/312/commits/41977050b94d1e1498a35ba3f7c393048f0da8d5.